### PR TITLE
Add export to scala client

### LINKED ISSUE #1567: implement export on top of Spark input format

### PR DESCRIPTION
**Pull request by itaiad200**
close #1567 

Still not done:
#1659 #1660 #1661

### CODE REVIEW

**Review by johnnyaug — APPROVED** [review]
Neat!
Approving, but see comments for your consideration

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package io.treeverse.clients
+
+import org.apache.spark.SerializableWritable
+import org.apache.spark.sql.{DataFrame, Row, SparkSession}
+
+class Exporter(apiClient: ApiClient, repoName: String, rootExportLocation: String) {
+  def exportAllFromBranch(spark : SparkSession, branch: String): Unit = {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala]: Moved the session and kept everything else.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      "LEFT OUTER JOIN new_commit " +
+      "ON (new_commit.key = prev_commit.key) " +
+      "WHERE new_commit.key is NULL")
+
+    export(spark, ns, rel, actionsDF)
+    spark.sparkContext.stop()
+  }
+}
+
+object RowHandler {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala]: Thanks, learned about *companion* objects.

File: clients/spark/examples/src/main/scala/io/treeverse/clients/examples/Export.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package io.treeverse.clients.examples
+
+import io.treeverse.clients.{ApiClient, Exporter}
+import org.apache.spark.sql.SparkSession
+
+object Export extends App {
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/examples/src/main/scala/io/treeverse/clients/examples/Export.scala:11]: Done.

File: clients/spark/examples/src/main/scala/io/treeverse/clients/examples/Export.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    val rootLocation = args(3)
+
+    val spark = SparkSession.builder().appName("I can list").master("local").getOrCreate()
+
+    val sc = spark.sparkContext
+    sc.hadoopConfiguration.set("lakefs.api.url", endpoint)
+    sc.hadoopConfiguration.set("lakefs.api.access_key", accessKey)
+    sc.hadoopConfiguration.set("lakefs.api.secret_key", secretKey)
+    sc.hadoopConfiguration.set("fs.s3a.access.key", s3accessKey)
+    sc.hadoopConfiguration.set("fs.s3a.secret.key",s3secretKey)
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/examples/src/main/scala/io/treeverse/clients/examples/Export.scala]: Done.

**Review by arielshaqed — COMMENTED** [review]
Sorry for late review.

I think we need a story about handling errors (_not_ in this PR, though!): when copying a million objects, one of them can go wrong.  How does the user discover what went wrong?

Other than that, important change to consider is to use `foreachPartition` and speed things up.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    // pin Exporter field to avoid serialization
+    val dst = dstRoot
+    val actionsDF = spark.sql("SELECT 'copy' as action, new_commit.* FROM new_commit " +
+      "LEFT JOIN prev_commit " +
+      "ON (new_commit.etag = prev_commit.etag AND new_commit.key = prev_commit.key) " +
+      "WHERE prev_commit.key is NULL " +
+      "UNION " +
+      "SELECT 'delete' as action, prev_commit.* FROM prev_commit " +
+      "LEFT OUTER JOIN new_commit " +
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:51]: Yes, this is messy, #1659 for the fix.

File: clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    export(ns, dst, actionsDF)
+    spark.sparkContext.stop()
+  }
+
+  private def export(ns: String, rel: String, actionsDF: DataFrame) =  {
+    val hadoopConf = spark.sparkContext.hadoopConfiguration
+    val serializedConf = new SerializableWritable(hadoopConf)
+
+    actionsDF.foreach { row =>
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply clients/spark/core/src/main/scala/io/treeverse/clients/Exporter.scala:29]: Exploring it, I don't think it's accurate. Found an [article](https://sparkbyexamples.com/spark/spark-foreachpartition-vs-foreach-explained/) about the difference, I don't see how this is related.

**Pull request merged**